astropy/table/operations.py

```diff
@@ -673,10 +606,22 @@ def _join(left, right, keys=None, join_type='inner',
 
         left_name, right_name = col_name_map[out_name]
         if left_name and right_name:  # this is a key which comes from left and right
-            out[out_name] = out.ColumnClass(length=n_out, name=out_name, dtype=dtype, shape=shape)
-            out[out_name] = np.where(right_mask,
```

Comment:
This was a latent bug: the previous statement (setting `out[out_name] = out.ColumnClass(..)` was actually being entirely ignored because this statement now makes a completely new column instead of in-place assignment.  Oops.